modules/gdpr_consent/gdpr_consent.module

```diff
@@ -0,0 +1,446 @@
+  }
+
+  $form['target_id'] = array(
+    '#type' => 'select',
+    '#title' => t('User consent agreement'),
+    '#default_value' => (isset($settings['target_id'])) ? $settings['target_id'] : '',
+    '#required' => FALSE,
+    '#options' => $agreements,
+    '#element_validate' => array(
+      'element_validate_integer_positive',
+    ),
+    '#description' => t('The GDPR User Consent Agreement to display'),
+  );
+
+  return $form;
+}
+
+/**
+ * Implements hook_field_formatter_view().
+ */
+function gdpr_consent_field_formatter_view($entity_type, $entity, $field, $instance, $langcode, $items, $display) {
+  $element = array();
+
+  switch ($display['type']) {
+    case 'gdpr_consent':
+      foreach ($items as $delta => $item) {
+        $element[$delta] = array(
+          '#type' => 'html_tag',
+          '#tag' => 'p',
+          '#value' => t('User Consent ID: @entity', array('@entity' => $item['target_id'])),
+        );
+      }
+      break;
+  }
+
+  return $element;
+}
+
+/**
+ * Implements hook_field_widget_info().
+ *
+ * Field widget to show consent information.
+ */
+function gdpr_consent_field_widget_info() {
+  return array(
+    'gdpr_consent_widget' => array(
+      'label' => t('GDPR Consent'),
+      'field types' => array('gdpr_user_consent'),
+    ),
+  );
+}
+
+/**
+ * Implements hook_field_widget_form().
+ */
+function gdpr_consent_field_widget_form(&$form, &$form_state, $field, $instance, $langcode, $items, $delta, $element) {
+  global $user;
+
+  $agreed = isset($items[$delta]['agreed']) ? $items[$delta]['agreed'] : '';
+  $notes = isset($items[$delta]['notes']) ? $items[$delta]['notes'] : '';
+
+  $widget = $element;
+  $widget['#delta'] = $delta;
+
+  // Get current revision of the referenced agreement entity.
+  $entity_type = 'gdpr_consent_agreement';
+  $entity = entity_load_single('gdpr_consent_agreement', $field['settings']['target_id']);
+
+  list($id, $vid, $bundle) = entity_extract_ids($entity_type, $entity);
+
+  switch ($instance['widget']['type']) {
+
+    case 'gdpr_consent_widget':
+      $element['agreed'] = array(
+        '#type' => 'checkbox',
+        '#title' => $entity->title,
+        '#default_value' => $agreed,
+        '#weight' => 0,
+        '#description' => $entity->long_description,
+      ) + $widget;
+      $element['notes'] = array(
+        '#type' => 'textfield',
+        '#title' => 'Consent Notes',
+        '#default_value' => $notes,
+        '#weight' => 10,
+        '#description' => '',
+        '#access' => user_access('grant any consent', $user),
+      ) + $widget;
+      $element['target_id'] = array(
+        '#type' => 'hidden',
+        '#value' => $field['settings']['target_id'],
+      ) + $widget;
+      $element['target_revision_id'] = array(
+        '#type' => 'hidden',
+        '#value' => $vid,
+      ) + $widget;
+      $element['date'] = array(
+        '#type' => 'hidden',
+        '#value' => REQUEST_TIME,
```

Comment:
Should this be captured on submit rather than form build? Should we do anything special if the revision id has changed since a prior commit?